Path: Integrate 4th-axis feature to PathPocketShape and 3D Surface ops

Russ4262 opened the pull request:
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [x] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---
 src/Mod/Path/Gui/Resources/panels/PageBaseGeometryEdit.ui
 shrink height of geometry list box
 
 src/Mod/Path/Gui/Resources/panels/PageOpPocketExtEdit.ui
 spelling correction

 src/Mod/Path/PathScripts/PathAreaOp.py
 integrate 4th-axis feature

 src/Mod/Path/PathScripts/PathPocketShape.py
 integrate 4th-axis feature

 src/Mod/Path/PathScripts/PathSurface.py
 integrate 4th-axis feature
 update existing OCL Dropcutter op
 update OCL Waterline op

 src/Mod/Path/Gui/ViewProviderPath.cpp
 correct B-axis rotational rendering for tool paths in PathWB

Each of these new features has been posted to the FreeCAD forum for feedback and testing.
3D Surface is discussed in [CNC Lathe and 4th-axis milling...](https://forum.freecadweb.org/viewtopic.php?f=15&t=26469), with script in PR posted at [this page](https://forum.freecadweb.org/viewtopic.php?f=15&t=26469&start=120#p299717).

Path Pocket Shape is discussed in [Path Pocket Shape: 4th-axis Integration](https://forum.freecadweb.org/viewtopic.php?f=15&t=35867), but began in "CNC Lathe and 4th-axis milling..." topic after the 3D Surface discussion.

Russ4262 commented:
The 3D Surface op is already in the experimental feature group. Therefore, suggested changes are still considered "experimental" also.  Both the Path Pocket Shape and 3D Surface scripts are posted to FreeCAD forums for testing and feedback.

luzpaz commented:
OP didn't link forum thread.  
JIC, CCing @mlampert @sliptonic

mlampert commented on src/Mod/Path/PathScripts/PathPocketShape.py:
```diff
@@ -204,104 +235,339 @@ def pocketInvertExtraOffset(self):
+                        tup = face, sub, angle, axis, tag, strDep, finDep, trans
+                        horizTuples.append(tup)
+            # Eol
+            '''
```
Personally I'm not very excited about commented out blocks of code (also applicable to other places) - can refactor this into it's own function and not call that, I think that would be much preferable.

Russ4262 replied:
I followed your suggestion, moving the code to a local function.  I then commented out the two lines calling the new local function.  In doing this, I realized a potential vertical faces grouping scenario that may now be unavailable to users.  I may need to hold this PR, and let you submit your extension changes, then rebase the 4th-axis implementation to the updated PathPocketShape.py file.  The scenario I am thinking of would only apply when "Use Rotation" is active, not "Off."  Your thoughts?

mlampert replied:
Not sure I understand the implications. Does it break existing functionality or does it represent a hole in the new feature? I would only hold the PR if it breaks existing functionality.

Russ4262 replied:
I think it will only be a hole in the new feature. I don't think existing features will be affected with "Use Rotation" set to default of "Off."  Set to "Off" I believe existing functionality is preserved.

Russ4262 commented on src/Mod/Path/PathScripts/PathAreaOp.py:
```diff
@@ -30,11 +36,18 @@
+__lastModified__ = "2019-04-29 10:43 CST"
+print("\n\nPathAreaOp.py")
+print(" -Script version: " + __scriptVersion__ + "  Lm: " + __lastModified__)
 
```
I made the conversions from print() to PathLog.debug().  I will still need to refine the contained messages and implement the translate() function within the PathLog.debug() calls.

Russ4262 commented on src/Mod/Path/PathScripts/PathPocketShape.py:
```diff
@@ -86,24 +101,31 @@ def selectOffsetWire(feature, wires):
+    except Exception as e:
+        msg = "\nThe selected face cannot be used.\nYou must select the bottom face of the pocket area.\nextendWire() in PathPocketShape.py"
+        FreeCAD.Console.PrintError(e)
+        FreeCAD.Console.PrintError(msg)
```
I made this change to use the PathLog.error() method.

Russ4262 commented:
@luzpaz 
I have added forum links to the PR description.  
Thank you for mentioning the links and interacting with this PR.

mlampert commented:
Right - it seems your changes break the unit tests, which is why the build fails. I forgot to tell you about that - sorry.

If you run the unit tests with 
`FreeCADCmd -t TestPathApp`

it'll tell you all about it.  What happened is that you added a new type into the system which we haven't dealt with before so all you need to do is add an entry in `PathSetupSheetOpPrototype` for each property type it doesn't know about yet. If you map them to `Property` it basically means its currently not supported (you'll know what to do if you have a look).

Russ4262 commented:
@mlampert :
I ran the test you mentioned, "FreeCADCmd -t TestPathApp" in my windows powershell. I redirected the output to a text file, tpa.txt.  I have attached that file. 
[tpa.txt](https://github.com/FreeCAD/FreeCAD/files/3131872/tpa.txt)

The only failure I see is due to an error in the dogbone dressup test as pasted here:
`ERROR: test02 (PathTests.TestPathDressupDogbone.TestDressupDogbone)
Verify bones are correctly generated for a Profile.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\rjohnson\Downloads\FreeCAD\Mod\Path\PathTests\TestPathDressupDogbone.py", line 135, in test02
    locs = sorted([bone[1] for bone in dog.bones], key=lambda xy: xy[0] * 1000 + xy[1])
AttributeError: ObjectDressup instance has no attribute 'bones'
`

As for the introduction of new properties, I see none. I looked through the initOperation() methods in each class within the PathAreaOp, PathPocketShape, and PathSurface files.  All properties I added fall within existing types, as I see them: PropertyEnumeration, PropertyBool, PropertyFloat, etc...

Am I missing something? 

I have clicked on the AppVeyor build checks, "Details" link for information about the failure.  Reading through some of the raw logs, I see the same failure reference to the dogbone dressup mentioned above.  I do not find any errors related to the PathAreaOp, PathPocketShape, and PathSurface script code I have modified for this PR.  I must be missing something.  

My apologies for the trouble, Gentlemen.

Russ4262 commented:
I may have found the issue. I read through the details of the latest Travis CI build report, raw version. I appears I had relocated the instantiation of self.depthparams in the opExecute() method of the PathAreaOp script. I reinstated the original instantiation in the original location: it will be updated as needed for the new feature.  I think this will fix at least one fail error.

Russ4262 commented:
@admins
Please hold this PR until some errors are corrected.  I edited the title to reflect this request. @sliptonic and @mlamper are assisting in identifying and correcting errors.
Thanks.

Russ4262 commented:
Hallelujah!  "All checks have passed" Finally...  Thanks to all of you working on the PCH and other issues related to appeasing the AppVeyor beast.  I will compile this repo from source locally and run the unit tests.  I have been having trouble building locally with the libpack on Windows using MSVSC 2017.  Will report back.

Russ4262 commented:
Well, @sgrogan and @mlampert , since I cannot get a successfully-compiled (100% success, 0 failed, 0 skipped) release version to actually run correctly on my Windows 10 machine, I shall transplant the changed python scripts in this PR into the pre-compiled  0.19_16616 binary and run the unit tests with it. If those tests are successful, I will remove the "[Hold]" tag in the title of this PR.

I will also post in the FC install/compile forum for assistance there - related to my inability to get a successful release (or debug) build to run correctly.

Thanks

OS: Windows 10 (10.0)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.16616 (Git)
Build type: Release
Branch: master
Hash: f94cdfd798d0c493efe9c7d45084ad2c5dd18caf
Python version: 3.6.8
Qt version: 5.12.1
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: English/United States (en_US)

mlampert commented:
I checked out your PR and ran it locally and all tests pass. It seems the build system got through it as well.

Russ4262 commented:
Thank you, Sir.  
Do I need to click the "Update branch" button still or leave this PR as is?
`This branch is out-of-date with the base branch
Merge the latest changes from master into this branch. `

mlampert commented on src/Mod/AddonManager/AddonManager.py:
```diff
@@ -586,19 +585,6 @@ def run(self):
-                    state = 0
-                else:
-                    state = 1
-                repos.append([name,url,state])
```
why is this section removed?

Russ4262 replied:
Sir, I did not edit this file.

mlampert commented on src/Mod/AddonManager/AddonManager.py:
```diff
@@ -339,7 +339,7 @@ def check_updates(self):
         self.repos.append(addon_repo)
         if addon_repo[2] == 1 :
-            self.listWorkbenches.addItem(QtGui.QListWidgetItem(QtGui.QIcon.fromTheme("dialog-ok",QtGui.QIcon(":/icons/edit_OK.svg")),str(addon_repo[0]) + str(" (Installed)")))
+            self.listWorkbenches.addItem(QtGui.QListWidgetItem(QtGui.QIcon.fromTheme("dialog-ok"),str(addon_repo[0]) + str(" (Installed)")))
```
why remove the icon?

Russ4262 replied:
I did not, Sir. I have not edited this file.

Russ4262 replied:
Fixed

mlampert commented on src/Mod/Sketcher/Gui/CommandSketcherTools.cpp:
```diff
@@ -1815,7 +1815,7 @@ CmdSketcherRectangularArray::CmdSketcherRectangularArray()
     sGroup          = QT_TR_NOOP("Sketcher");
     sMenuText       = QT_TR_NOOP("Rectangular Array");
-    sToolTipText    = QT_TR_NOOP("Creates a rectangular array pattern of the geometry taking as reference the last selected point");
+    sToolTipText    = QT_TR_NOOP("Creates an rectangular array pattern of the geometry taking as reference the last selected point");
```
'not a native english speaker but I could swear it's supposed to be `a`

Russ4262 replied:
Indeed it should. I never changed this file, Sir.

mlampert commented on src/Mod/Path/Gui/Resources/panels/PageOpPocketExtEdit.ui:
```diff
@@ -107,7 +107,7 @@
           <widget class="QPushButton" name="buttonDisable">
            <property name="toolTip">
-            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Diisable the currently selected pocket extension.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Disable the currently selected pocket extension.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Thanks for fixing this.

Russ4262 replied:
No problem, mate. Was browsing and it caught my eye.

Russ4262 replied:
Fixed

mlampert commented on src/Gui/ApplicationPy.cpp:
```diff
@@ -1267,39 +1267,20 @@ PyObject* Application::sGetMarkerIndex(PyObject * /*self*/, PyObject *args)
-        if (std::find(std::begin(sizeList), std::end(sizeList), defSize) == std::end(sizeList))
-            defSize = 9;
-
-        return Py_BuildValue("i", Gui::Inventor::MarkerBitmaps::getMarkerIndex(marker_arg, defSize));
```
The changes here look like a merge/rebase issue - wanna check out the latest version from `master` and replace this one?

Russ4262 replied:
@mlampert 
>The changes here look like a merge/rebase issue - wanna check out the latest version from master and replace this one?

Sir, I have never changed: 
*src/Mod/AddonManager/AddonManager.py
*src/Mod/Sketcher/Gui/CommandSketcherTools.cpp 
*src/Mod/Sketcher/Gui/CommandSketcherTools.cpp 
*src/Gui/ApplicationPy.cpp 

If I understand you correctly, you would like me to copy each of the files mentioned FROM current `master` to my fork `master` so these issues will be resolved?

Russ4262 commented:
@mlampert 
The `gitlog.log` file referenced states it was deleted.  It doesn't show up on my repo/fork.  May I assume it is a temporary file?  If I clicke 'Files changed' at the top of this PR page, the `gitlog.log` says it has been deleted.  Should I select 'delete file' from the sub-menu associated with the file? Doing so would fix that issue, correct?

Otherwise, I fixed the other non-personal PR files affected by the improper rebase issue you mentioned above.  The `gitlog.log` file issue is the only one remaining.

mlampert commented:
yeah I don't think the gitlog file was supposed to get committed, that's fine.

I wasn't sure about the other files, but it seems there might've been a merge issue which is why those changes crept in. It happens. I'll have another look.

Russ4262 commented:
@mlampert
Should I delete the gitlog.log file in the "files changed" section of this PR so it doesn't cause a problem? Or should I leave it alone?

I apologize for the rebase error issue. Thanks for helping me fix it, Sir.

mlampert commented:
I think you can leave it as is.

There is absolutely no need to apologize for anything here. These things happen and that is one reason why we have code reviews - I've messed up way worse in the past.

Russ4262 commented:
Issue raised by RatonLaveur on FC forum: [Issue with operation Depths on modified path scripts](https://forum.freecadweb.org/viewtopic.php?f=15&t=36166).  Multi-face selection was broken in current scripts. As of this comment the issue has been identified and corrected within the PathPocketShape.py script.  I will push fixed script to my fork to update this PR.  Thanks gentlemen.
